clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java

```diff
@@ -239,8 +239,10 @@ void run(long now) {
      * Start closing the sender (won't actually complete until all data is sent out)
      */
     public void initiateClose() {
-        this.running = false;
+        // It is important to close the record accumulator before setting the running flag to false. Otherwise we may
```

Comment:
Maybe a comment like this would be clearer:
```java
// ensure accumulator is closed first to guarantee that no more appends are accepted after
// breaking from the sender loop. Otherwise, we may miss some callbacks when shutting down.
```